gio/gdocumentportal.c

```diff
@@ -0,0 +1,126 @@
+/* GIO - GLib Input, Output and Streaming Library
+ *
+ * Copyright 2016 Endless Mobile, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General
+ * Public License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <errno.h>
+
+#include "gdocumentportal.h"
+#include "xdp-dbus.h"
+#include "gstdio.h"
+
+#ifdef G_OS_UNIX
+#include "gunixfdlist.h"
+#endif
+
+static GXdpDocuments *documents;
+static char *documents_mountpoint;
+
+static void
+init_document_portal (void)
+{
+  static gsize documents_inited = 0;
+
+  if (g_once_init_enter (&documents_inited))
+    {
+      GError *error = NULL;
+      GDBusConnection *connection = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, &error);
+
+      if (connection != NULL)
+        {
+          documents = gxdp_documents_proxy_new_sync (connection, 0,
+                                                     "org.freedesktop.portal.Documents",
+                                                     "/org/freedesktop/portal/documents",
+                                                     NULL, NULL);
+          gxdp_documents_call_get_mount_point_sync (documents,
+                                                    &documents_mountpoint,
+                                                    NULL, NULL);
+
+          g_object_unref (connection);
+        }
+      else
+        {
+          g_warning ("Cannot connect to session bus when initializing document portal: %s",
+                     error->message);
+          g_error_free (error);
+        }
+
+      g_once_init_leave (&documents_inited, 1);
+    }
+}
+
+char *
+g_document_portal_add_document (GFile   *file,
+                                GError **error)
+{
+  char *doc_path, *path, *basename;
+  char *doc_id = NULL;
+  char *doc_uri = NULL;
+  GUnixFDList *fd_list = NULL;
+  int fd, fd_in;
+  gboolean ret;
+
+  init_document_portal ();
+
```

Comment:
Logic after this point does not make sense if, for any reason, we don't have `documents` or `documents_mountpoint` properly initialized. Maybe print a warning, set the error and exit would be a good idea in that case.